mpir: introduce hardware topology abstraction layer

gcongiu opened the pull request:
Current `comm_split_type` implementation relies on `hwloc`/`netloc` direct
calls, which have to be guarded by ifdefs to allow builds that do not
use them to work. However, this solution is not ideal and makes the code
hard to maintain and prone to bugs.

This patch introduces a new hardware topology abstraction layer that can
be used by the MPIR layer. It allows to remove all the direct references
to `hwloc`/`netloc` currently present in the MPIR layer and move all the
hardware topology specific code under a simple, clean API.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,189 @@
+#define MPIR_HW_TOPO_H_INCLUDED
+
+/*
+ * Tracks hwloc_obj_type_t do not change ordering
```
Why is the ordering important?

gcongiu replied:
I just wanted to have the possibility to pass these directly to `hwloc` without checking if `MPIR_HWLOC_OBJ_MACHINE` is actually equal to `HWLOC_OBJ_MACHINE` and doing the conversion if not.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,189 @@
+    MPIR_HWLOC_OBJ_TYPE_MAX,
+    MPIR_MPICH_OBJ_DDR,         /* MPICH object for DDR memory */
+    MPIR_MPICH_OBJ_MCDRAM,      /* MPICH object for MCDRAM memory */
+    MPIR_MPICH_OBJ_DEFAULT = MPIR_MPICH_OBJ_DDR
```
1. I don't think we should use `HWLOC` or `MPICH` in the name.  Just `MPIR_NODE_HW_OBJ_` might be sufficient.

2. If the enum type is `MPIR_node_hw_obj_type`, then the values should be `MPIR_NODE_HW_OBJ_TYPE_*`.

3. I think we are moving to the naming convention `MPIR_NODE_HW_OBJ_TYPE__MACHINE` (i.e., two underscores before the actual type).

4. I don't think `MPIR_NODE_HW_OBJ_TYPE__DEFAULT` should be a part of the enum itself.  It should be an external global variable (preferably as a part of the `MPIR_Process` structure).  Bonus points if it can be a CVAR-updatable value.

gcongiu replied:
All these sound reasonable I will address them. I like the `CVAR` option for the default.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,189 @@
+    MPIR_HWLOC_MEMBIND_INTERLEAVE = 3,
+    MPIR_HWLOC_MEMBIND_NEXTTOUCH = 4,
+    MPIR_HWLOC_MEMBIND_MIXED = -1
+} MPIR_Hwloc_membind_policy;
```
How about `MPIR_MEMBIND_POLICY__*`?  Also, by `default`, do you mean system-specified?

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,189 @@
+    MPIR_HWLOC_MEMBIND_MIGRATE = (1 << 3),
+    MPIR_HWLOC_MEMBIND_NOCPUBIND = (1 << 4),
+    MPIR_HWLOC_MEMBIND_BYNODESET = (1 << 5)
+} MPIR_Hwloc_membind_flags;
```
How about `MPIR_MEMBIND_FLAGS__*`?

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,189 @@
+    MPIR_MPICH_NETLOC_OBJ_NETWORK,
+    MPIR_MPICH_NETLOC_OBJ_TOPOLOGY,
+    MPIR_MPICH_NETLOC_OBJ_NODE
+} MPIR_Netloc_obj_class;
```
How about `MPIR_NETWORK_HW_OBJ_TYPE__*`?

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,189 @@
+    MPIR_NETLOC_NETWORK_TYPE_ETHERNET = 1,
+    MPIR_NETLOC_NETWORK_TYPE_INFINIBAND = 2,
+    MPIR_NETLOC_NETWORK_TYPE_INVALID = 3
+} MPIR_Netloc_network_type;
```
How about `MPIR_NETWORK_TYPE__*`?

gcongiu replied:
Actually after looking better at the code this type of information can be hidden in `topo_util.c` since it is needed only by `netloc` itself which will not be exposed directly here. What we probably want to expose to `MPIR` is just something like:

```
typedef enum {
    MPIR_NETWORK_QUERY_BY__SWITH_LEVEL,
    MPIR_NETWORK_QUERY_BY__MIN_SIZE,
    MPIR_NETWORK_QUERY_BY__MIN_MEMSIZE,
    MPIR_NETWORK_QUERY_BY__TORUS_DIMENSION
} MPIR_Network_query_by;

int (*get_covering_netobj) (MPIR_Network_query_by key, int val, int *out);
```

Where `val` can be either `switch_level`, `subcomm_min_size`, `min_mem_size` or `dimension`, respectively. Am I interpreting this correctly?

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,189 @@
+typedef enum {
+    MPIR_NETLOC_TOPOLOGY_TYPE_INVALID = -1,
+    MPIR_NETLOC_TOPOLOGY_TYPE_TREE = 1
+} MPIR_Netloc_topology_type;
```
How about `MPIR_NETWORK_TOPOLOGY_TYPE__*`?

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,189 @@
+    MPIR_NETLOC_NODE_TYPE_HOST = 0,
+    MPIR_NETLOC_NODE_TYPE_SWITCH = 1,
+    MPIR_NETLOC_NODE_TYPE_INVALID = 2
+} MPIR_Netloc_node_type;
```
How about `MPIR_NETWORK_NODE_TYPE__*`?

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,189 @@
+    MPIR_MPICH_NETLOC__SUBCOMM_MIN_SIZE,
+    MPIR_MPICH_NETLOC__MIN_MEM_SIZE,
+    MPIR_MPICH_NETLOC__DIMENSION
+} MPIR_Netloc_search_key;
```
These are not specific to netloc or the network.  They are just logical partitioning models.  How about `MPIR_COMM_SPLIT_MODEL__*`?

pavanbalaji commented:
Apart from my above review comments, a more general comment -- I was hoping that all these will be abstracted in their own directory instead of being inside the `init` directory.

gcongiu commented:
@pavanbalaji I have updated the code. Hopefully this version is better

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,152 @@
+    MPIR_NODE_HW_OBJ__NETWORK_DEVICE = 19,
+    MPIR_NODE_HW_OBJ__DDR = 20,
+    MPIR_NODE_HW_OBJ__HBM = 21,
+} MPIR_Node_hw_obj_type;
```
I think the objects should be `MPIR_NODE_HW_OBJ_TYPE__*` to be consistent with the name of the enum.

gcongiu replied:
Sorry my mistake

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,152 @@
+    MPIR_MEMBIND_POLICY__INTERLEAVE = 3,
+    MPIR_MEMBIND_POLICY__NEXTTOUCH = 4,
+    MPIR_MEMBIND_POLICY__MIXED = -1
+} MPIR_Membind_policy;
```
`MPIR_MEMBIND_POLICY__DEFAULT` should be outside the enum as a constant.  Also bonus points if it's a CVAR.

pavanbalaji replied:
Actually, I had a previous discussion point for this -- do you really mean `default` or `system-specified`?

gcongiu replied:
If the user says `default` there will be a system defined default not user defined. So it should not be a CVAR (unlike for the memory type). E.g., for linux it will be `firsttouch`, for solaris `bind`. Unless we want to change the behavior, in which case we have to expose this to be set by the user.

pavanbalaji replied:
I'd recommend calling it `SYSTEM`, instead of `DEFAULT`.

How does the user control which of these policies we'd use?  Isn't that a CVAR?  Should it be?

gcongiu replied:
These will be available to users through CVARs yes. Each buffer will have a `mem_type`, `mem_policy` and `mem_flags` (you can look at PR #3200 to see how they are defined for each buffer). This will be integrated in the other heterogeneous memory PR though. Later, if we find it useful, we can add more buffers.

pavanbalaji replied:
Since the memory binding pieces are not yet in `master`, it might be better to move this part to the other PR that introduces memory binding.

gcongiu replied:
Ok

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,152 @@
+    MPIR_NETWORK_QUERY_TYPE__MIN_SIZE,
+    MPIR_NETWORK_QUERY_TYPE__MIN_MEMSIZE,
+    MPIR_NETWORK_QUERY_TYPE__TORUS_DIMENSION
+} MPIR_Network_query_type;
```
Some of these are not network query types, e.g., `MIN_SIZE` and `MIN_MEMSIZE`.  Those are logical queries.

gcongiu replied:
Maybe I should find a better naming. These will be selecting type of queries for `netloc` so I don't have to provide three separate interfaces.

pavanbalaji replied:
To be clear: they are not actual queries to `netloc`.  For example, if the `MIN_SIZE` can fit inside a node, you don't need to query netloc at all.  Similarly, even if netloc is not available, these queries can be supported.  They are not directly connected to the physical topology, although their behavior would be better with physical topology information.

`TORUS_DIMENSION`, on the other hand, is directly related to network topology.

gcongiu replied:
I need to think better about how to split these to make the API as concise as possible. It seems putting this all together in the same interface will just cause confusion.

pavanbalaji replied:
I agree.

pavanbalaji commented on src/util/hw_topo/mpir_hw_topo.c:
```diff
@@ -0,0 +1,168 @@
+static int MPIR_GEN_get_covering_nodeobj(MPIR_Node_hw_obj_type obj_type, int *size, int **obj_ids);
+static int MPIR_GEN_mbind(void *addr, size_t len, MPIR_Node_hw_obj_type mem_obj_type, int *obj_ids,
+                          MPIR_Membind_policy policy, MPIR_Membind_flags flags);
+#endif
```
Can the function name be kept the same and two different implementations be provided with or without hwloc?

gcongiu replied:
Sure, just wanted to have a distinction between the names to separate the two. So in principle you could, theoretically, have another topology backend (e.g., `MPIR_XXX_`) and still have a general common function to fall back to if neither `MPIR_HWLOC` or `MPIR_XXX` is available. But I guess since we don't have additional hardware topology libraries support this is superfluous.

pavanbalaji replied:
I understand, but I guess my point was to have just one function name, which would map to different backends depend on what we configure mpich with.

gcongiu commented:
@pavanbalaji I have removed the reference to memory topology and gave up on the idea of exposing only a simple function that would return the color to `comm_split_type` for the queried network parameters. It's easier, and faster, to leave the `comm_split_type` code there and just get rid of direct references to implementation specific network topology parameters, replacing them with function calls.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,184 @@
+ * Forward declarations for network objects
+ */
+#define MPIR_Network_node void
+#define MPIR_Network_edge void
```
Can you use a `typdef` instead of a define?

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,184 @@
+     *
+     * TODO: add detailed description
+     */
+     MPIR_Network_type(*network_get_type) ();
```
You should add a `void` parameter to the function to avoid future warnings.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,184 @@
+     *
+     * TODO: add detailed description
+     */
+    MPIR_Network_node *(*network_get_endpoint) ();
```
`void` parameter needed.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,184 @@
+     *
+     * TODO: add detailed description
+     */
+    int (*network_get_num_nodes) ();
```
`void` parameter.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,184 @@
+     *
+     * TODO: add detailed description
+     */
+    int (*network_torus_get_dimension) ();
```
`void` parameter.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,184 @@
+     *
+     * TODO: add detailed description
+     */
+    int *(*network_torus_get_geometry) ();
```
`void` parameter.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,184 @@
+     *
+     * TODO: add detailed description
+     */
+    int (*network_torus_get_node_index) ();
```
`void` parameter.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,184 @@
+     *
+     * TODO: add detailed description
+     */
+    int *(*network_tree_get_node_levels) ();
```
`void` parameter.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,184 @@
+     */
+    int (*network_tree_get_switches_at_level) (int switch_level,
+                                               MPIR_Network_node *** switches_at_level,
+                                               int *switch_count);
```
Functions specific to a topology should be marked as such inside a separate structure.  So, `tree` functions should go in a structure.  `torus` functions should go in a different structure.  These structures can be in a union.  That way, it's clear that only one set is valid at a time.

gcongiu replied:
Doesn't that have a lot of levels of dereference? Following this logic I should probably have a `node_struct` for node functions and then a `network_struct` for network functions. Then inside `network_struct` I can have the `union`. Thus accessing topology function would look like `hw_topo_fns->node->get_processor_obj`, `hw_topo_fns->network->get_type` and `hw_topo_fns->network->u.tree->get_switches_at_level`.

pavanbalaji replied:
Structures are not dereferences.  `hw_topo_fns->node.get_processor_obj`, `hw_topo_fns->network.u.tree.get_switches_at_level`.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,184 @@
+/*
+ * Initialize hardware topology
+ */
+int MPIR_Hardware_topology_init();
```
`void` parameter.

pavanbalaji commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,184 @@
+/*
+ * Finalize hardware topology
+ */
+int MPIR_Hardware_topology_finalize();
```
`void` parameter.

pavanbalaji commented on src/util/hw_topo/mpir_hw_topo.c:
```diff
@@ -0,0 +1,289 @@
+/*
+ * Hardwre topology attributes
+ */
+static struct MPIR_Hardware_topology {
```
If this is an internal structure, it shouldn't be namespaced.

pavanbalaji commented on src/util/hw_topo/mpir_hw_topo.c:
```diff
@@ -0,0 +1,289 @@
+static int MPIR_Network_tree_get_hostnode_index(int *node_index, int *num_nodes);
+static int MPIR_Network_tree_get_switches_at_level(int switch_level,
+                                                   MPIR_Network_node *** switches_at_level,
+                                                   int *switch_count);
```
No namespace for local functions.

pavanbalaji commented on src/util/hw_topo/mpir_hw_topo.c:
```diff
@@ -0,0 +1,289 @@
+#define FUNCNAME MPIR_Network_get_type
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+static MPIR_Network_type MPIR_Network_get_type()
```
`void` parameter.

pavanbalaji commented on src/util/hw_topo/mpir_hw_topo.c:
```diff
@@ -0,0 +1,289 @@
+#define FCNAME MPL_QUOTE(FUNCNAME)
+static MPIR_Network_node *MPIR_Network_get_endpoint()
+{
+
```
Weird extra line.

gcongiu commented:
test:jenksin/ch3/tcp
test:jenkins/ch4/ofi

gcongiu commented:
test:jenkins/ch3/tcp
test:jenkins/ch4/ofi

gcongiu commented:
@pavanbalaji tests seem fine. I have changed slightly the naming convention I was using earlier but I can easily move back if you prefer. There might also be some more tewaking to be done in `comm_split_type`. Unfortunately, I haven't had the time to simplify the `comm_split_type` code. So I am wrapping the needed `hwloc` functions and adding a few more to access information from opaque objects (e.g., `MPIR_Node_obj`). Hopefully this is close to the final version.

pavanbalaji commented:
@gcongiu I was hoping for a slightly higher-level abstraction.  For example, why does the comm_split_type code need to know about bitmaps at all?  All it really needs to know if the index of a certain type of object that it resides on.  So I was hoping for something like:

```
obj = get_parent_obj_of_type(type);
idx = get_obj_index(obj);
```

If `type` is `SOCKET`, for example, the `get_parent_obj_of_type` function would return this process' parent object that is of type `SOCKET`.  If the process' bitmask spans more than one socket, the function would return `NULL` or something similar.  That way, the bitmasks are all hidden inside these functions.

gcongiu commented:
@pavanbalaji the shared memory part is not difficult to implement that way, although I will still need some extra function for device queries, but the network part uses bitmaps all over the place. bitmaps are also sent across to other processes and then used for further queries. For this reason I though I would just do a wrapper for the needed `hwloc` functions and leave the MPIR layer the flexibility to use them. I can reorganize things better but I will need more time.

pavanbalaji commented:
In what cases do network queries need bitmaps?  I'd need a device ID which I can send out, but I don't know why bitmaps need to be sent.

gcongiu commented:
@pavanbalaji when splitting the communicator by subcomm minsize if there are more processes in the node than the requested size it continues by further splitting inside the node. To do this it `1.` gets the lowest depth, in the topology tree, of the covering object of every process in the node (e.g., `L3CACHE`), `2.` goes through all the objects at that level and checks what processes are covered by that particular object cpuset and `3.` uses this info to build a bitmap (more an array of integers) of processes (`process_cpuset[i]`). This bitmap, along with the position of the current process in it is used to get the color. The step `2.` requires bitmap operations because the `cpuset` of the object at the `lowest_depth` can have more bits set (more `PU`s). This is what it looks like. I will see if I can get rid of the bitmap operations.

gcongiu commented:
test:jenkins/ch3/tcp

gcongiu commented:
test:jenkins/ch3/tcp

gcongiu commented:
@pavanbalaji I have updated this PR to include bug fixes https://github.com/pmodels/mpich/pull/3636 and https://github.com/pmodels/mpich/pull/3628 and move a little towards the final version. Hopefully the current topology interface is sufficiently high level.

gcongiu commented:
test:jenkins/ch3/tcp

gcongiu commented:
test:jenkins/ch3/tcp

gcongiu commented:
test:jenkins/ch3/tcp

gcongiu commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,225 @@
+    MPIR_NODE_OBJ_OSDEV_TYPE__DMA,
+    MPIR_NODE_OBJ_OSDEV_TYPE__COPROC
+} MPIR_Node_obj_osdev_type;
+
```
@pavanbalaji I left this unchanged as I wasn't sure renaming it as `MPIR_Node_osdev_type` would be the best thing to do. This is currently following the `hwloc` nomenclature. In fact all objects, regardless of their type, are represented as `MPIR_Node_obj` (`hwloc_obj_t`). Maybe an alternative naming could be `MPIR_Node_osdev_obj_type`?

pavanbalaji replied:
I take back my comment about this change.  On reading it again, the current naming does make sense.  `MPIR_Node_obj_type` defines object types.  When we pick `MPIR_NODE_OBJ_TYPE__OS_DEVICE`, the `MPIR_Node_obj_osdev_type` enum gives more information on which OS device.  So you can leave it as it is.

gcongiu commented:
@pavanbalaji I noticed today there is a PR that was merged on May 1st that introduces topology-aware intranode trees (supported through `hwloc`). This PR cannot be merged unless that code is modified to account for the new interface.

gcongiu commented:
@jain-surabhi-23 can you have a look at the last patch? The only change I made is in how the `bind_map` is built. I have run `allred` test locally using the configuration information you provided and it seems to work.

jain-surabhi-23 commented:
@gcongiu The patch looks fine just by looking at it. But I will have to make sure the topology tree is same as we expected. Other thing I will make sure is this code gets built on MacOS correctly. That is why I had to guard the code with macro you are have taken out in that commit.

gcongiu commented:
@jain-surabhi-23 actually I just noticed for some reason compilation was not catching a left over reference to `hwloc` in the topology code. I will fix it and let you know when this can be reviewed

gcongiu commented:
@jain-surabhi-23 I have added another patch to fix the aforementioned problem. Hopefully this should work.

gcongiu commented:
@jain-surabhi-23 I have rebased on latest master, can you review the last patch?

jain-surabhi-23 commented:
@gcongiu I will review this today. Looks good but I would want to run it before approving.

jain-surabhi-23 commented:
@gcongiu This PR seems to have broken the debug information we print for intra-node topology aware trees. If you set `MPIDI_SHM_TOPOTREE_DEBUG` to `1` in `shm/src/topotree_types.h` and run this -
`MPIR_CVAR_REDUCE_POSIX_INTRA_ALGORITHM=release_gather MPIR_CVAR_BCAST_POSIX_INTRA_ALGORITHM=release_gather mpirun -n 10 -bind-to core ./bcast_comm_world_only -type=MPI_INT -count=1 -seed=9 -testsize=1024`, there is a segmentation fault. It does not happen with master.
Also remember to configure with "izem=atomic"

gcongiu commented:
@jain-surabhi-23 thank you, I will look into the problem

gcongiu commented:
@pavanbalaji I have refactored Intels's `topotree` code to use the new topology infrastructure. Could you look at this PR one more time and see if there is anything missing? I would like to merge this ASAP as it is also needed by other open PRs

gcongiu commented:
test:jenkins/ch3/tcp
test:jenkins/ch4/ofi

gcongiu commented:
test:mpich/warnings/more

gcongiu commented:
test:mpich/warnings/more

gcongiu commented:
test:mpich/warnings/more

gcongiu commented:
test:jenkins/ch3/tcp

gcongiu commented:
test:mpich/warnings/more
test:jenkins/ch3/tcp

gcongiu commented:
test:mpich/warnings/more
test:jenkins/ch3/tcp

gcongiu commented:
test:mpich/warnings/more

gcongiu commented:
test:mpich/warnings/more

gcongiu commented:
test:mpich/warnings/more

gcongiu commented:
test:mpich/warnings

gcongiu commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

gcongiu commented:
test:mpich/warnings/more

gcongiu commented:
@pavanbalaji there are some leftover warnings that came up because previously guarded parts of the `comm_split_type` implementation are not guarded anymore. I will fix those, meanwhile can you look at the code? I need this PR to be merged in order to work on the heterogeneous memory integration needed for the November's release.

gcongiu commented:
test:mpich/warnings/more

gcongiu commented:
test:mpich/warnings/more

gcongiu commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

gcongiu commented:
@hzhou warnings disappeared. Thanks for the hints!

hzhou commented on src/util/hw_topo/mpir_hw_topo.c:
```diff
@@ -0,0 +1,948 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
```
It is just a single file, so let's flatten it to `src/util/mpir_hw_topo.c`.

gcongiu replied:
That directory also contains `mpir_netloc.c`

hzhou replied:
let's move both

hzhou commented on src/util/hw_topo/mpir_hw_topo.c:
```diff
@@ -0,0 +1,948 @@
+enum {
+    TO,
+    FROM
+};
```
Even though this enum are only used locally, it still covers the scope of entire file, and make it a bit more verbose, e.g. add some prefix, may help the readability.

gcongiu replied:
These are only used as arguments within conversion functions, is it really necessary to prefix them?

hzhou replied:
It is not really necessary, of course. But simple words should be avoided unless it is only used in a narrow scope when reviewers can see its context in one view.  Here, what does `TO/FROM` mean? Adding a few prefix or comment to it so one does't need search them to figure out their purposes would be nice.

gcongiu replied:
These directions are used with, e.g., `obj_type_to_from_hw_obj_type` and `TO/FROM` select the direction from which the conversion happens. `TO` means I give the function a `MPIR_NODE_OBJ_TYPE__` and I get a corresponding hardware object type `HWLOC_OBJ_TYPE_`. From means the opposite, I give the function an hwloc object type and I get an MPIR object type in return.

hzhou replied:
I see. How about `MPIR_OBJ_TO_HWLOC` and `HWLOC_OBJ_TO_MPIR`?

hzhou replied:
Instead of `obj_type_to_from_hw_obj_type`, how about `convert_hw_obj_type`?

gcongiu replied:
mmm this name seems to implicitly say that the function always takes an hw object type and converts it into something else. The initial name I was using was `mpir_obj_type_to_from_hw_obj_type` but that `mpir` prefix seemed like a namespace for a static function. `convert_obj_type` might be better as you can say what object type you want to convert using `MPIR_OBJ_TO_HWLOC` and `HWLOC_OBJ_TO_MPIR`?

hzhou replied:
convert _node_obj_type?

gcongiu replied:
`convert_node_obj_type`, `convert_node_os_obj_type` and `convert_net_obj_type`?

hzhou replied:
sounds good

hzhou commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,232 @@
+} MPIR_Node_obj_type;
+
+/*
+ * Node hardware OS device object types
```
I would suggest add note that these are sub-types for `MPIR_NODE_OBJ_TYPE__OS_DEVICE` -- unless I interpreted wrong.

hzhou commented on src/include/mpir_hw_topo.h:
```diff
@@ -0,0 +1,232 @@
+ */
+typedef void MPIR_Network_node;
+typedef void MPIR_Network_edge;
+
```
Just a question: are we constructing a graph here, that node is node and edge is edge, and `MPIR_Node_obj` are the attribute/value of the node?

gcongiu replied:
`MPIR_Node_obj` are node local objects that map to `hwloc`. `MPIR_Network_node/edge` are network objects that map to `netloc`. We construct network topology (graph) using `MPIR_Netloc_` functions in `MPIR_Hardware_topology_init`. When upper layers query the topology layer (either node or network) they get an opaque handle which they can use to get additional information from the topology layer. This way we don't need to expose any topology internal information to them. For example, you may want to get the id of the memory device covering the PU your process is running to bind memory to that device. In this case you query the topology layer which will return the opaque `MPIR_Node_obj` handle to the memory device; afterwards you can further query the the memory `id` by passing the such handle to another function.

hzhou replied:
I see. Thanks.

pavanbalaji replied:
Why are these `void` instead of `void *`?  How can someone represent anything with a `void`?

gcongiu replied:
If you prefer I can define these as `void *` and change all the definitions from`MPIR_Node_obj *obj` to `MPIR_Node_obj obj`.

pavanbalaji replied:
I'm not sure that's a preference issue.  It's a correctness issue.  You can't represent anything with just `void`.  So you are working around the issue by using pointers in the calling functions.  For example, functions that need to return a node object, now have to return a pointer to a node object.

gcongiu replied:
I see, I will fix it

gcongiu replied:
Fixed

gcongiu replied:
If there is nothing else I missed I will rerun tests one more time

hzhou commented on src/mpi/init/init_topo.c:
```diff
@@ -29,50 +29,10 @@
-        }
-    }
-#endif
+    MPIR_Hardware_topology_init();
```
You may as well remove `init_topo.c` and directly call `MPIR_Hardware_topology_...()` directly. `init_topo.c` was just a placeholder when I did the refactor before a more thorough investigation.

gcongiu replied:
Right, I forgot about that one, I will add another commit for that

gcongiu commented:
test:mpich/warnings/more
test:jenkins/ch3/most
test:jenkins/ch4/most

gcongiu commented:
test:mpich/warnings/more
test:jenkins/ch3/most
test:jenkins/ch4/most

gcongiu commented:
@pavanbalaji @hzhou tests seem fine but for solaris which for some, apparently unrelated reason, is failing.

hzhou commented:
> @pavanbalaji @hzhou tests seem fine but for solaris which, for some apparently unrelated reason, is failing.

It timed out after two hours. @raffenet just added a 2-hour total time-out limit on the review jobs. Apparently, they are not sufficient for `solaris-ch3`

raffenet commented:
test:jenkins/ch3/most

gcongiu commented:
@raffenet `ch3/tcp` completes now, `ch3/sock` still terminates before finishing

hzhou commented:
> @raffenet `ch3/tcp` completes now, `ch3/sock` still terminates before finishing

Because I extended the job timeout from 120min to 150 min this morning, but I neglected to do the same for `sock`. It is also updated now.

gcongiu commented:
test:jenkins/ch3/sock

gcongiu commented:
@pavanbalaji I have rearranged the commits and made some additional change to address Hui's comments. Tests are passing. Is there anything else I should fix?

gcongiu commented:
Running again:
test:mpich/warnings/more
test:jenkins/ch3/most
test:jenkins/ch4/most

gcongiu commented:
test:mpich/warnings/more
test:jenkins/ch3/most
test:jenkins/ch4/most

gcongiu commented:
@pavanbalaji can you have another look at this PR? Since last review I have fixed the `typedef void *MPIR_Node_obj` and split the `convert_` functions in `mpir_hw_topo.c` into two, one to convert from `MPIR_Node_obj_type` to `hwloc_obj_type_t` and one for doing the opposite conversion. This solves some warning that were detected when compiling the code with `--enable-strict`.

pavanbalaji commented on src/util/mpir_hw_topo.c:
```diff
@@ -0,0 +1,564 @@
+enum {
+    MPIR_OBJ_TO_HW,
+    HW_OBJ_TO_MPIR
+};
```
Weird naming.  `MPIR_` is a prefix.  What is this enum for?

gcongiu replied:
It should not be there anymore. I thought I had removed it.

pavanbalaji commented:
Some of the functions are named `MPIR_`.  Is there a reason for naming them `MPIR_` instead of `MPII_`?

gcongiu commented:
> Some of the functions are named `MPIR_`. Is there a reason for naming them `MPIR_` instead of `MPII_`?

You mean `MPIR_Hardware_topology_init/finalize`?

gcongiu commented:
test:mpich/warnings/more

gcongiu commented:
test:jenkins/ch4/ofi

gcongiu commented:
Rerun after some cleanup:
test:mpich/warnings/more
test:jenkins/ch3/most
test:jenkins/ch4/most

hzhou commented:
test:more/warnings

hzhou commented:
test:more/warnings

hzhou commented:
The pgi compiler failures should go away after rebase to master

pavanbalaji commented:
Approved this PR, with the assumptions that the tests will pass.

gcongiu commented:
Rerunning after rebase onto master
test:jenkins/ch3/most
test:jenkins/ch4/most
